pkg/server/endpoints/node/limiter.go

```diff
@@ -0,0 +1,185 @@
+package node
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	"golang.org/x/time/rate"
+	"google.golang.org/grpc/peer"
+)
+
+const (
+	// Max burst size
+	attestLimit int = 1
+	csrLimit    int = 500
+	jsrLimit    int = 500
+
+	AttestMsg = iota
+	CSRMsg
+	JSRMsg
+)
+
+type Limiter interface {
+	Limit(ctx context.Context, msgType, count int) error
+}
+
+// Newlimiter returns a new node api rate.Limiter
+func NewLimiter(l logrus.FieldLogger) *limiter {
+	return &limiter{
+		attestRate:   1,
+		csrRate:      500,
+		jsrRate:      500,
+		lastNotified: make(map[string]time.Time),
+		limiters:     make(map[int]map[string]*rate.Limiter),
+		log:          l,
+	}
+}
+
+type limiter struct {
+	// Allowed number of messages per second
+	attestRate rate.Limit
+	csrRate    rate.Limit
+	jsrRate    rate.Limit
+
+	lastNotified map[string]time.Time
+	limiters     map[int]map[string]*rate.Limiter
+	log          logrus.FieldLogger
+	mtx          sync.Mutex
+}
+
+// Limit enforces rate limiting policy by blocking until the specified number of messages can
+// be processed. It introspects the context in order to identify the caller. An error will be
+// returned if the context is cancelled, an invalid msgType is specified, or if the number
+// of messages exceeds the burst limit.
+func (l *limiter) Limit(ctx context.Context, msgType, count int) error {
```

Comment:
so if i understand correctly, the limits are imposed per caller, per msg type? 500 ops a second per caller seems a tad high....

Reply:
That's correct, though the "limit" is the max burst size - no request with more than 500 CSRs will be allowed to pass. The actual rate is defined when the struct is created, currently capped at 100 per second... In this example, if a request was submitted with 500 CSRs, it would (under the worst conditions) have to wait 5 seconds for it to be permitted.

Since exceeding this limit would be a permanent error condition, I have taught the agent to truncate requests that would exceed it, allowing the remainder to be submitted on the next manager cycle